File: stackusingqueue.py
# Time Complexity : for all functions O(1) because there's not search or complex thing like that
# Space Complexity : I dont know how to calculate space complexity, here i assume it's O(2n) because we have
# two queues, then again i feel like it's O(1) for no reason. 
# Did this code successfully run on Leetcode :
#  Any problem you faced while coding this :
class MyQueue(object):

    def __init__(self):
        self.inQueue = []
        self.outQueue = []

    def push(self, x):
        self.inQueue.append(x)   

    def pop(self):
        self.peek()
        return self.outQueue.pop()

    def peek(self):
        if not self.outQueue:
            while(self.inQueue):
                self.outQueue.append(self.inQueue.pop())
            else:
                print("main stack is empty")
        return self.outQueue[-1]

    def empty(self):
        return not self.inQueue and not self.outQueue

# Your MyQueue object will be instantiated and called as such:
# obj = MyQueue()
# obj.push(x)
# param_2 = obj.pop()
# param_3 = obj.peek()
# param_4 = obj.empty()

File: StackUsingQueue.java
// Time Complexity : O(1)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No
import java.util.Stack;

class StackUsingQueue {
    private Stack<Integer> inStack;
    private Stack<Integer> outStack;

    public StackUsingQueue() {
        this.inStack = new Stack<>();
        this.outStack = new Stack<>();
    }

    public void push(int x) {
        inStack.push(x);
    }

    public int pop() {
        if(outStack.isEmpty()) {
            while(!inStack.isEmpty()) {
                outStack.push(inStack.pop());
            }
        }
        return outStack.pop();
    }

    public int peek() {
        if(outStack.isEmpty()) {
            while(!inStack.isEmpty()) {
                outStack.push(inStack.pop());
            }
        }
        return outStack.peek();
    }

    public boolean empty() {
        return inStack.isEmpty() && outStack.isEmpty();
    }
}

/**
 * Your MyQueue object will be instantiated and called as such:
 * MyQueue obj = new MyQueue();
 * obj.push(x);
 * int param_2 = obj.pop();
 * int param_3 = obj.peek();
 * boolean param_4 = obj.empty();
 */
